BudgetSquirrel.Api/Services/Implementations/AsyncQueryService.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+using BudgetSquirrel.Business;
+using BudgetSquirrel.Data.EntityFramework;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query;
+
+namespace BudgetSquirrel.Api.Services.Implementations
+{
+  public class AsyncQueryService : IAsyncQueryService
+  {
+    private readonly BudgetSquirrelContext context;
+
+    public AsyncQueryService(BudgetSquirrelContext context)
+    {
+      this.context = context;
+    }
+
+    public IQueryable<T> Include<T, TProperty>(IQueryable<T> source, Expression<Func<T, TProperty>> include) where T : class
+    {
+      if (source is DbSet<T>)
+      {
+        DbSet<T> includable = ((DbSet<T>) source);
+        return includable.Include(include);
+      }
+      else
+      {
+        throw new InvalidOperationException("source must be a DbSet, not " + source.GetType().Name);
+      }
+    }
+
+    public Task<T> SingleOrDefaultAsync<T>(IQueryable<T> source, Expression<Func<T, bool>> predicate)
+    {
+      return source.SingleOrDefaultAsync(predicate);
```

Comment:
Why would we not just call this instead of our own version. 

Also why do we not await it if it is an Async method.

Reply:
> Why would we not just call this instead of our own version.

Because the domain layer doesn't have a reference to entity framework core. I can add a document on that. This is basically a shim to provide an interface to the domain layer to use entity core methods without actually needing entity framework. It's messy and it's one of the biggest beefs with EF core but it's better than using non-async calls or including EFCore in the domain layer.

> Also why do we not await it if it is an Async method.

Always prefer returning the task directly if you don't need to use the result in that method. There is a lot of overhead with awaiting a task. using await actually still does this exact same thing under the hood, but it returns a wrapper task around the task it's awaiting... That duplication of tasks around tasks causes a lot of processor overhead, making the method slower.